## LangChain

In [1]:
%pip install -q python-dotenv langchain-openai


[notice] A new release of pip is available: 24.0 -> 26.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [2]:
from dotenv import load_dotenv

load_dotenv()

True

In [3]:
from langchain_openai import ChatOpenAI

query = "P&G 회사의 it 직무는 어떤걸 하나요?"

llm = ChatOpenAI(model="gpt-4o-mini")

llm.invoke(query)

AIMessage(content='P&G(프로터 앤 갬블)의 IT 직무는 다양하며, 회사의 기술적 요구와 비즈니스 목표를 지원하는 역할을 수행합니다. 일반적으로 P&G의 IT 직무에서 담당하는 업무는 다음과 같습니다:\n\n1. **시스템 개발 및 유지보수**: 기업의 IT 시스템을 설계하고 개발하며, 기존 시스템의 유지보수를 담당합니다. 이를 통해 효율성을 높이고 비즈니스 프로세스를 최적화합니다.\n\n2. **데이터 분석**: 빅데이터와 분석 도구를 활용하여 시장 동향, 소비자 행동 등을 분석하고, 이를 통해 경영진이 전략적 의사 결정을 내릴 수 있도록 지원합니다.\n\n3. **IT 지원 및 서비스 운영**: 직원들이 사용하는 IT 시스템과 애플리케이션에 대한 기술 지원을 제공합니다. 문제 해결, 기술 교육 및 시스템 최적화를 포함합니다.\n\n4. **보안 관리**: 회사의 데이터와 시스템을 보호하기 위한 사이버 보안 전략을 수립하고 이행합니다. 보안 위협에 대비하고, 데이터 유출 방지 조치를 마련합니다.\n\n5. **프로젝트 관리**: IT 관련 프로젝트를 계획하고 관리하여, 예산 내에서 정해진 기한 내에 프로젝트를 완수하도록 합니다. 다양한 부서와 협업하여 목표를 달성합니다.\n\n6. **신기술 도입**: 인공지능(AI), 클라우드 컴퓨팅, IoT(사물인터넷) 등 최신 기술을 연구하고 도입하여 비즈니스 혁신을 추진합니다.\n\nP&G의 IT 직무는 글로벌 기업의 특성상 다양한 문화와 환경에서 일하게 되며, 팀워크와 커뮤니케이션 능력도 중요합니다. IT는 회사의 전체 비즈니스 전략과 밀접하게 연결되어 있기 때문에, 해당 직무는 회사의 성과에 큰 영향을 미칠 수 있습니다.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 408, 'prompt_tokens': 20, 'total_tokens': 428, 'completion_tok

AIMessage(content='P&G(프로터 앤 갬블)의 IT 직무는 다양하며, 회사의 기술적 요구와 비즈니스 목표를 지원하는 역할을 수행합니다. 일반적으로 P&G의 IT 직무에서 담당하는 업무는 다음과 같습니다:\n\n1. **시스템 개발 및 유지보수**: 기업의 IT 시스템을 설계하고 개발하며, 기존 시스템의 유지보수를 담당합니다. 이를 통해 효율성을 높이고 비즈니스 프로세스를 최적화합니다.\n\n2. **데이터 분석**: 빅데이터와 분석 도구를 활용하여 시장 동향, 소비자 행동 등을 분석하고, 이를 통해 경영진이 전략적 의사 결정을 내릴 수 있도록 지원합니다.\n\n3. **IT 지원 및 서비스 운영**: 직원들이 사용하는 IT 시스템과 애플리케이션에 대한 기술 지원을 제공합니다. 문제 해결, 기술 교육 및 시스템 최적화를 포함합니다.\n\n4. **보안 관리**: 회사의 데이터와 시스템을 보호하기 위한 사이버 보안 전략을 수립하고 이행합니다. 보안 위협에 대비하고, 데이터 유출 방지 조치를 마련합니다.\n\n5. **프로젝트 관리**: IT 관련 프로젝트를 계획하고 관리하여, 예산 내에서 정해진 기한 내에 프로젝트를 완수하도록 합니다. 다양한 부서와 협업하여 목표를 달성합니다.\n\n6. **신기술 도입**: 인공지능(AI), 클라우드 컴퓨팅, IoT(사물인터넷) 등 최신 기술을 연구하고 도입하여 비즈니스 혁신을 추진합니다.\n\nP&G의 IT 직무는 글로벌 기업의 특성상 다양한 문화와 환경에서 일하게 되며, 팀워크와 커뮤니케이션 능력도 중요합니다. IT는 회사의 전체 비즈니스 전략과 밀접하게 연결되어 있기 때문에, 해당 직무는 회사의 성과에 큰 영향을 미칠 수 있습니다.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 408, 'prompt_tokens': 20, 'total_tokens': 428, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-mini-2024-07-18', 'system_fingerprint': 'fp_e220a5ab5b', 'id': 'chatcmpl-DaxVMpQ7NHTW3eWjaxk7cmfLmi3Mc', 'service_tier': 'default', 'finish_reason': 'stop', 'logprobs': None}, id='lc_run--019de73b-45fd-7492-85ee-13155188e545-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 20, 'output_tokens': 408, 'total_tokens': 428, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}})

## LangGraph
- State, Node, Edge, Conditional Edge

In [ ]:
# State : 상태(현재 Agent의 상태)
# Node : Agent가 수행하는 task (Node는 다른 Agent가 될 수도 있고, 함수이다)